File: Exams/2014/Solutions/2014-07-15/4/CPP/Chess/ChessPiece/ChessPiece.cpp
#include "ChessPiece.h"

ChessPosition CHESS_PIECE::ChessPiece::getPosition() {
    return currentPosition;
}

bool CHESS_PIECE::ChessPiece::captures(ChessPosition pos) {
    vector<ChessPosition> moves = allowedMoves();

    return any_of(moves.begin(), moves.end(), [&](ChessPosition p) { return p == pos; });
}

File: past exams/2014/Solutions/2014-07-15/4/Java/Chess/src/com/fmi/chess/piece/ChessPiece.java
package com.fmi.chess.piece;

import com.fmi.chess.position.ChessPosition;

import java.util.List;

public abstract class ChessPiece {
    protected ChessPosition currentPosition;

    public ChessPosition getPosition() {
        return currentPosition;
    }

    public void setCurrentPosition(ChessPosition currentPosition) {
        this.currentPosition = currentPosition;
    }

    public abstract List<ChessPosition> allowedMoves();

    public abstract PieceType getType();

    public Boolean captures(ChessPosition pos) {
        return allowedMoves().stream().anyMatch(p -> p.equals(pos));
    }
}
